File: Lecture24/shoot.py
import math
import cpt
import bvp


    
print " Steping solution of u'' = - pi^2 (u+1) / 4"
x0, x1, u0, u1 = input( "Input x0, x1, u0, u1 : ")
delta = input( "Input initial guess : ")

stepper = bvp.BVPStep( x0=x0, x1=x1, u0=u0, u1=u1, delta=delta, N= 100)
shooter = bvp.BVPShoot( step=stepper, x0=x0, x1=x1, u0=u0, u1=u1, delta=delta, N= 100)

slope = shooter.solve()
print ''
print " Shooting method found slope = ", slope
print ""
print ""
x_plot = []
u_plot = []
#print " Trajectory : "
stepper.xy = [ x0, u0, slope ]
#print '{0:>8s} {1:>8s} {2:>8s}'.format( 'x', 'u', "u'" )
#print '{0:8.4f} {1:8.4f} {2:8.4f}'.format( stepper.xy[0], stepper.xy[1], stepper.xy[2] )
x_plot.append( stepper.xy[0] )
u_plot.append( stepper.xy[1] )
for i in range(stepper.N):
    stepper.step()
    #print '{0:8.4f} {1:8.4f} {2:8.4f}'.format( stepper.xy[0], stepper.xy[1], stepper.xy[2] )
    x_plot.append( stepper.xy[0] )
    u_plot.append( stepper.xy[1] )


import matplotlib
matplotlib.rcParams['legend.fancybox'] = True
import matplotlib.pyplot as plt

plt.plot( x_plot, u_plot )

plt.show()

File: Lecture24/shoot.cpp
#include "bvp.hpp"
using namespace cpt;

int main() {
  BVPStep bvpStepper;
  BVPShoot bvpShooter( bvpStepper );
    cout << " Shooting solution of u'' = - pi^2 (u+1) / 4\n"
         << " with boundary conditions u(0) = 0, u(1) = 1\n";

    cout << " using initial guess delta = " << bvpShooter.get_delta() << endl;
    double delta = bvpShooter.solve();
    cout << "\n Secant search found delta = " << delta << endl;

    bvpShooter.set_initial_values();           // to integrate the final solution
    bvpStepper.update_values( bvpShooter );    // Initialize to the solved equation
    ofstream data_file("shoot.data");

    Matrix<double,1> p = bvpStepper.get_p();

    data_file << p[BVP::_x] << '\t' << p[BVP::_u] << '\t' << p[BVP::_dudx] << '\n';

    for (int i = 0; i < bvpShooter.get_N(); i++) {
      bvpStepper.step(); 
      p = bvpStepper.get_p();
      data_file << p[BVP::_x] << '\t' << p[BVP::_u] << '\t' << p[BVP::_dudx] << '\n';
    }
    data_file.close();
    cout << "\n " << bvpStepper.get_N() << " step solution in file shoot.data"
         << endl;
}
